File: search-2d-matrix/javascript/solution.js
/**
 * @param {number[][]} matrix
 * @param {number} target
 * @return {boolean}
 */
var searchMatrix = function (matrix, target) {
  let hiRow = matrix.length;
  let loRow = 0;

  do {
    let row = Math.floor(loRow + (hiRow - loRow) / 2);

    let hiColumn = matrix[row].length;
    let loColumn = 0;

    do {
      let column = Math.floor(loColumn + (hiColumn - loColumn) / 2);

      if (matrix[row][column] === target) {
        return true;
      } else if (matrix[row][column] < target) {
        loRow = row + 1;
        loColumn = column + 1;
      } else {
        hiRow = row;
        hiColumn = column;
      }
    } while (loColumn < hiColumn);
  } while (loRow < hiRow);
  return false;
};

console.log(
  searchMatrix(
    [
      [1, 3, 5, 7],
      [10, 11, 16, 20],
      [23, 30, 34, 60],
    ],
    3
  )
);
// expect true
console.log(
  searchMatrix(
    [
      [1, 3, 5, 7],
      [10, 11, 16, 20],
      [23, 30, 34, 60],
    ],
    13
  )
);
// expect false

File: search-2d-matrix/go/solution.go
package main

func searchMatrix(matrix [][]int, target int) bool {
	var loRow, hiRow int = 0, len(matrix)
	var curRow int = loRow + (hiRow-loRow)/2
	for loRow < hiRow {
		var loCol, hiCol = 0, len(matrix[0])
		var curCol int = loCol + (hiCol-loCol)/2

		for loCol < hiCol {
			if matrix[curRow][curCol] == target {
				return true
			}

			if matrix[curRow][curCol] < target {
				loRow = curRow + 1
				loCol = curCol + 1
			} else {
				hiRow = curRow
				hiCol = curCol
			}
			curCol = loCol + (hiCol-loCol)/2
		}
		curRow = loRow + (hiRow-loRow)/2
	}
	return false
}
